### ISSUE TITLE
Cron Job Takes Ages

### ISSUE THREAD

**Issue by MparkG** [nextcloud/maps#294]
Hello
is there a way to debug the cron job (cron.php)?
Id like to know where it is stuck so long.
I had to turn off the cronjob because the processes kept adding up and did not finish in 5 minutes.

Thanks

**Comment by MparkG** [comment]
dont have logs, or any errors to report in the admin protocoll log list.
that said, heres what i have done lately
i have, for example, external folders with a shitton of family pictures from everyone, so thats quite a number of 
pics.
Now, 
ive installed that plugin, just curious to see what it does.
Next thing i see are emails from cronjobs with errors like "php ran out of memory, 521mb allowed, trying for 10mb"
Tendencially a kinda popular yet error that pops up often, and usually not because of memory limitations.
Cronjob hammered out one more process of "php cron.php" every 5min.
However the processes persisted, as each one takes like half an hour.
So ever a while a few of them were running simultanuously.

So to fix this, i thought maybe it executed some one-time-job for another external folder ive set up at about the same time.
So i ran this filesystem:scan or similar task, to index all files.. 
Meanwhile the oc_jobs table contained this 68000-somewhat entries of the aforementioned task.

Ive got a user account for each family member and that external folder is shared amongst us, everyone has access.
(For the pic preview and file scanning task, this is kind of annoying as it appearantly goes through all files not on a file basis, but on a user basis. So it takes quite long, reading in a shared folder 5 times, once for each user...)

These tasks in oc_jobs are, i assume, also somewhat like 5 times the file count in the pictures folder.
I wouldnt believe to have 70k Pics, no way, but something like 20k could be well possible.

Then i removed the plugin again and cleaned the jobs table.

**Comment by e-alfred** [comment]
Maybe you have issues with this bug: https://github.com/nextcloud/maps/pull/251

**Comment by bascht** [comment]
I think I ran into the same issue here:

![Screenshot-2020-02-19-112526](https://user-images.githubusercontent.com/51264/74848282-d0d2bd00-5337-11ea-837d-26a4440c53ba.png)

Can I safely delete all maps jobs e.g. `DELETE * FROM oc_jobs WHERE class LIKE "OCA\\Maps%"; ?

**Comment by bascht** [comment]
Update: Deleted all the oc_jobs for Maps, re-ran the occ scan (took about 7 hours) and now everything seems fine. I think my issue was also related to #251.

**Comment by bascht** [comment]
Update: The Hoards of idle CronJobs are back – I'm just gonna uninstall Maps for now. :worried: 
![Screenshot-2020-02-25-085858](https://user-images.githubusercontent.com/51264/75226746-29c5a980-57ad-11ea-973d-0946600a847c.png)

**Comment by tacruc** [comment]
Well the occ scan will create a job for each picture. Thats what it is supposed to do.
In 0.1.4 there was a Problem that each job failed. So only one row was removed every 15 minutes, which is a problem. Now it should process ruffly 1 row per second or 1000 rows per 15 min, which means your pictures are scaned the next few days. There is nothing to be imrpved, as reading the geo information just takes time.

**Comment by bascht** [comment]
@tacruc mhh, okay – but still weird though. After my first comment I manually triggered the cron via CLI until the whole job list was processed. Everything looked fine for the next days until I checked and the table was back at > 100k jobs for OCA\\Maps. That's why I commented again. :-)

**Comment by tacruc** [comment]
But less. The cron job should be executed every 15 min. If this is fine, the background Jobs should be done in a finte amount of time. 
It is not supposed to be finished by a single call to cron.php.
Cron.PHP will process Jobs frome the Jobs table for a const amount of time (I think 15min, but the value does not matter) and then terminated. Then New cronjob has to be started to continue.

**Comment by bascht** [comment]
@tacruc Oh yeah, sorry – I should've been more clear. What I did last week:

* I deleted all the OCA\\Maps Jobs from the table
* Executed `php occ maps:scan-photos <my-user>` to re-enlist all the photos (table was at ~ 60k entries then)
* Processed the jobs with `while true; do php cron.php; done` until all the `OCA\\Maps\` jobs  where gone (that took about ~ 7 hours)

That was last week and it looked fine after processing the backlog. Photos showed up on the map.

On Monday I checked why my external calendars didn't get updated (basically the `OCA\\DAV\\BackgroundJob\\RefreshWebcalJob` weren't getting executed) I took another look at the `oc_jobs` table and it was back at > 100k entries.

**Comment by tacruc** [comment]
And you did not execute maps:scan-photos again?

Is the cron.php executed on a regular basis? What kind of cron are you using?

**Comment by bascht** [comment]
@tacruc Yip, I ran `maps:scan-photos` only once. `cron.php` is triggered via user crontab every 5 minutes.

**Comment by mduller** [comment]
> 
> 
> But less. The cron job should be executed every 15 min. If this is fine, the background Jobs should be done in a finte amount of time.
> It is not supposed to be finished by a single call to cron.php.
> Cron.PHP will process Jobs frome the Jobs table for a const amount of time (I think 15min, but the value does not matter) and then terminated. Then New cronjob has to be started to continue.

I ran into the same issue as described in this thread (350k entries in oc_jobs for Maps photo update in my case). They mostly kept failing with out of memory e-mails but I also saw how the **second cron job would be launched while the previous was still running**. That resulted in 2 postmaster (PostgreSQL) processes running at ~100%. 15 minutes later a third one was added. I then killed the 3 php processes driving the cron job and the postmaster processes returned to normal. I then cleaned the oc_jobs table from all Maps photo update jobs and also disabled the plugin for now.

**Comment by chaosgrid** [comment]
Hello there, we just had the same problem. I just took down our instance because the cron job was running at 100% and added 350k+ OCA\Maps\BackgroundJob\AddPhotoJob entries to our oc_jobs before I took it down!

How does this work? Does it scan all photos of every user? Because for large instances this is not doable at all (imagine millions of photos, no way a cron job can do that in a single run).

**Comment by chaosgrid** [comment]
Btw. here are two links to people with the same problem:

https://help.nextcloud.com/t/solved-cron-job-hang-permanent-with-no-errors-maps-app-problem/71325/2

https://help.nextcloud.com/t/background-jobs-never-end-high-cpu-mysql-after-update-17-0-3-18-0-1/72644

**Comment by ttbek** [comment]
Why does a scan take a whole second?  That means mine would take more than 6 days.  I can't imagine that the actual job takes that long right, more of the start/end overhead?  Maybe they could be done in batches of 50 or something?

**Comment by tacruc** [comment]
The pictue has to be loaded. Depending on your camera this is roughly 10mb.  I'm not saying that it takes a hole second but that it is in the order of a second. It might be that it processes 3-5 pictures a second or even up to 10. So the second is a guess. Nothing that I measured. 
I just wanted to say that scanning all pictures on a large insurance will never be fast. 
> How does this work? Does it scan all photos of every user? Because for large instances this is not doable at all (imagine millions of photos, no way a cron job can do that in a single run).

But the cronjob is designed to handle long running jobs. 
The cronjob will stop executing jobs after 15 minutes and the next cronjob will continue.
Yeah it will take some days, but it's only done once and in the background so it shouldn't influence the users experience.

**Comment by mduller** [comment]
@tacruc what you describe about the behavior of the cronjob would indeed be tolerable. However, at least for me this was not the behavior I was able to observe. I have seen a new cronjob added every 15 minutes (see my [earlier post](https://github.com/nextcloud/maps/issues/294#issuecomment-595568396)). This is, of course, a problem. Also, the high memory consumption was surprising: it exceeded 1GB and thus was aborted. I did not investigate the root cause but am suspecting that it was trying to read large video files into memory.

**Comment by tacruc** [comment]
I'm not sure how to continue with this issue, as I think there are many different problems. First of all, the task itself is not trivial/fast so it will take time.
* In the past when this issue was created there were additional issues. Which caused the cronjob to crash. But his issues are fixed.
* Apparently it can happen that the cronjons accumulate, which is a different problem. I would suggest to create a new issue for that, as it sounds more like a problem in organising the background jobs and not of the individual job. But we should investigate it.
* There might be an issue that background jobs are created faster than they are processed. I would suggest to open an issue for this.

I would close this issue now as it is quite general. And the first explanation is scanning millions of pictures just takes time. Still I would appreciate to the other issues discussed here, such that we have the place to investigate them appropriate.

**Comment by e-alfred** [comment]
@tacruc Nextcloud nowadays recommends running the Cronjob every 5 minutes, maybe this is a problem with accumulating Cronjobs? Manual for 18 here:

https://docs.nextcloud.com/server/18/admin_manual/configuration_server/background_jobs_configuration.html?highlight=cron#cron

**Comment by tacruc** [comment]
If the cronjob is limited to ~15 minutes there should be at most 4 cronjons running or? But I would like to end this discussion here and move it to the corresponding issues. As different aspects getting mixed in this issue. The original problem was that cronjob did not finish within 5 minutes. Which is fine.

**Comment by ttbek** [comment]
> The pictue has to be loaded. Depending on your camera this is roughly 10mb.

Umm, hmm, more like 50 in my case, but I guess I don't know why it needs to be loaded.  

> There is nothing to be imrpved, as reading the geo information just takes time.

I thought Maps just needed to pull metadata (GPS data from EXIF) as Nextcloud already generates thumbnails.  I have no idea what the current implementation is though, I haven't looked at the maps code.  
If the scan really does need to load the image (maybe you generate smaller thumbnails than the usual Nextcloud ones or something?), then yeah, makes sense.  If it isn't necessary, then EXIF scanning would probably get us to hundreds a second or even faster.

**Comment by tacruc** [comment]
So we use two lib's to read the exif, as both are not 100% perfect. But from my observations, both of them load the full file, or at least a hughe part of it. Implementing the exif loading ourself is unfortunately out of scope. 
@ttbek I'm happy to use any function you provide, which takes a path as argument and returns the GPS data and is faster and more stable than the current implementation. But unless you can't provide this function all considerations of how this can be done more efficiently are of theoretical nature.

**Comment by ttbek** [comment]
Have you tried exiftool?  
I can see it extracts GPS data like: 

> GPS Altitude                    : 6.3 m Below Sea Level
> GPS Date/Time                   : 2020:04:17 17:14:13Z
> GPS Latitude                    : 25 deg 19' 8.59" N
> GPS Longitude                   : 51 deg 29' 27.77" E
> GPS Position                    : 25 deg 19' 8.59" N, 51 deg 29' 27.77" E

I tried in a directory of mine that is about 50:50 SRW and JPG files.  
>  /usr/bin/time exiftool ./*
I omit all the EXIF written to screen here.
>     1 directories scanned
>  2383 image files read
> 33.29user 2.12system 1:42.52elapsed 34%CPU (0avgtext+0avgdata 26844maxresident)k
> 7861128inputs+0outputs (0major+101329minor)pagefaults 0swaps

So that's 23.24 images/second, but... I'm writing out the EXIF to the terminal, so lets both reduce that since writing to screen is slow, but also add the overhead of some grep query

> /usr/bin/time exiftool ./* | grep alsdkfkljhdfgkjshdfgkljsdhfg
> 22.58user 0.37system 0:22.96elapsed 99%CPU (0avgtext+0avgdata 27144maxresident)k
> 0inputs+0outputs (0major+112034minor)pagefaults 0swaps

So now we're getting 103.79 images per second.  
These are all NX1 images, the raw SRW ones are 40-50mb, there is no way exiftool is loading those entirely into memory in that time.  

exiftool is a well established tool for EXIF data, been around since 2003.  It is cross platform and written with Perl.  There are gps examples in the FAQ: https://exiftool.org/faq.html  

Main page is here: https://exiftool.org/
The security section is worth considering if it would be in Maps, might want to call it with some input sanitation for files starting with '-'.  

Code is here: https://github.com/exiftool/exiftool

exiftool can be given the file or a path or even a directory as an argument.  The internal gps selection shown in the FAQ should be faster than grep filtering for it afterwards.  

I might be misunderstanding the issue, sorry for wasting time if I am.
